libs/MeshKernel/src/CasulliDeRefinement.cpp

```
@@ -0,0 +1,1020 @@
+            }
+
+            // deactivate cell
+            mesh.m_numFacesNodes[connectedElementId] = 0;
+        }
+        else
+        {
+
+            // polygons of degree higher than three: remove node and link
+
+            for (UInt j = 0; j < mesh.m_numFacesNodes[connectedElementId]; ++j)
+            {
+                UInt edgeId = mesh.m_facesEdges[connectedElementId][j];
+
+                if (mesh.m_edgesNumFaces[edgeId] < 2)
+                {
+                    continue;
+                }
+
+                if (mesh.m_edgesFaces[edgeId][0] == elementId || mesh.m_edgesFaces[edgeId][1] == elementId)
+                {
+                    UInt ndum = mesh.m_numFacesNodes[connectedElementId] - 1;
+
+                    std::shift_left(mesh.m_facesEdges[connectedElementId].begin() + j, mesh.m_facesEdges[connectedElementId].begin() + ndum, 1);
+
+                    // remove one node per removed link
+                    // take the first node that has not been removed before, but not the node that is kept,
+                    // which is the first of the center cell
+
+                    UInt i = 0;
+
+                    while (i < mesh.m_numFacesNodes[connectedElementId] &&
+                           mesh.m_facesNodes[connectedElementId][i] != mesh.GetEdge(edgeId).first &&
+                           mesh.m_facesNodes[connectedElementId][i] != mesh.GetEdge(edgeId).second &&
+                           mesh.m_facesNodes[connectedElementId][i] != mesh.m_facesNodes[elementId][0])
+                    {
+                        ++i;
+                    }
+
+                    if (k < mesh.m_numFacesNodes[connectedElementId])
+                    {
+                        std::shift_left(mesh.m_facesNodes[connectedElementId].begin() + i, mesh.m_facesNodes[connectedElementId].begin() + ndum, 1);
+                    }
+                    else
+                    {
+                        // Error: No node found
```

do we need to throw an exception here?